src/tests/zfs/zfs_custom_node_topology/zfs_custom_node_topology_test.go

```diff
@@ -0,0 +1,397 @@
+
+	workerNodes, err := zfs.ListZfsNode(common.NSOpenEBS())
+	Expect(err).ToNot(HaveOccurred(), "failed to list worker node")
+
+	// minimum worker nodes in cluster should be two
+	Expect(len(workerNodes)).Should(BeNumerically(">=", 2),
+		"test case requires are least 2 worker nodes, %d nodes found", len(workerNodes))
+
+	key = "zfse2e/nodename"
+	targetNode = workerNodes[0]
+
+	// label worker node
+	err = k8stest.LabelNode(targetNode, key, targetNode)
+	Expect(err).ToNot(HaveOccurred(), "failed to label node %s", targetNode)
+
+	zfsScOptions := k8stest.ZfsOptions{
+		// PoolName:      nodeConfig.PoolName,
+		PoolName:      "zfspv-pool",
+		ThinProvision: common.No,
+		RecordSize:    "128k",
+		Compression:   common.Off,
+		DedUp:         common.Off,
+		AllowedTopologies: []coreV1.TopologySelectorTerm{
+			{
+				MatchLabelExpressions: []coreV1.TopologySelectorLabelRequirement{
+					{
+						Key:    key,
+						Values: []string{targetNode},
+					},
+				},
+			},
+		},
+	}
+
+	for i := 0; i <= len(workerNodes); i++ {
+		app := k8stest.FioApplication{
+			Decor:                   fmt.Sprintf("%s-%d", decor, i),
+			VolSizeMb:               1024,
+			OpenEbsEngine:           engine,
+			VolType:                 volType,
+			FsType:                  fstype,
+			Loops:                   1,
+			VolWaitForFirstConsumer: volBindModeWait,
+			Zfs:                     zfsScOptions,
+		}
+		appInstances = append(appInstances, &app)
+	}
+
+	// deploy all fio application
+	for ix, app := range appInstances {
+		logf.Log.Info(fmt.Sprintf("%d)", ix), "fio-pod", app.Decor, "err", err)
+		err = app.DeployApplication()
+		Expect(err).ToNot(HaveOccurred(), "failed to deploy %s, %v", app.Decor, err)
+	}
+
+	// verify all fio application are deployed on same labeled node
+	for _, app := range appInstances {
+		appPodName := app.GetPodName()
+		logf.Log.Info("app pod", "fio-pod", appPodName)
+		//get node name where where app is deployed
+		node, err := k8stest.GetNodeForPodByPrefix(appPodName, common.NSDefault)
+		Expect(err).ToNot(HaveOccurred(), "failed to node for app pod %s, %v", appPodName, err)
+		Expect(node).Should(Equal(targetNode), "app pod %s does not scheduled on node %s", targetNode)
+	}
+
+	// remove all fio application
+	for ix, app := range appInstances {
+		logf.Log.Info(fmt.Sprintf("%d)", ix), "fio-pod", app.Decor, "err", err)
+		// remove app pod, pvc,sc
+		err = app.Cleanup()
+		Expect(err).To(BeNil(), "failed to clean resources")
+	}
+
+	// Remove the labels from nodes after the end of test
+	err = k8stest.UnlabelNode(targetNode, key)
+	Expect(err).ToNot(HaveOccurred(), "failed to remove label from node %s", targetNode)
+	targetNode = ""
+
+	productConfig := e2e_config.GetConfig().Product
+	label := fmt.Sprintf("%s=%s", productConfig.LocalEngineComponentPodLabelKey,
+		productConfig.ZfsEngineComponentDsPodLabelValue)
+
+	// restart zfs daemonset pods
```

Comment:
Why the restart of daemonset pods?

Reply:
topology key present in csinode kubernetes object for local.csi.openebs.io plugin driver should be removed before starting new topology test and to do so , daemonset pods needs to restated after removing node label with the key

Reply:
Please add comments to code describing why and what. Thanks

Reply:
at all the places where we delete the pods.